cmd/enumerate_github/main.go

```diff
@@ -0,0 +1,238 @@
+	minStarsFlag        = flag.Int("min-stars", 10, "only enumerates repositories with this or more of stars.")
+	starOverlapFlag     = flag.Int("star-overlap", 5, "the number of stars to overlap between queries.")
+	requireMinStarsFlag = flag.Bool("require-min-stars", false, "abort if -min-stars can't be reached during enumeration.")
+	queryFlag           = flag.String("query", "is:public", "sets the base query to use for enumeration.")
+	workersFlag         = flag.Int("workers", 1, "the total number of concurrent workers to use.")
+	startDateFlag       = dateFlag(epochDate)
+	endDateFlag         = dateFlag(time.Now().UTC().Truncate(oneDay))
+	logFlag             = logLevelFlag(log.DefaultLevel)
+)
+
+// dateFlag implements the flag.Value interface to simplify the input and validation of
+// dates from the command line.
+type dateFlag time.Time
+
+func (d *dateFlag) Set(value string) error {
+	t, err := time.Parse(githubDateFormat, value)
+	if err != nil {
+		return err
+	}
+	*d = dateFlag(t)
+	return nil
+}
+
+func (d *dateFlag) String() string {
+	return (*time.Time)(d).Format(githubDateFormat)
+}
+
+func (d *dateFlag) Time() time.Time {
+	return (time.Time)(*d)
+}
+
+// logLevelFlag implements the flag.Value interface to simplify the input and validation
+// of the current log level.
+type logLevelFlag log.Level
+
+func (l *logLevelFlag) Set(value string) error {
+	*l = logLevelFlag(log.ParseLevel(value))
+	return nil
+}
+
+func (l logLevelFlag) String() string {
+	return string(log.Level(l))
+}
+
+func (l logLevelFlag) Level() log.Level {
+	return log.Level(l)
+}
+
+func init() {
+	flag.Var(&startDateFlag, "start", "the start `date` to enumerate back to. Must be at or after 2008-01-01.")
+	flag.Var(&endDateFlag, "end", "the end `date` to enumerate from.")
+	flag.Var(&logFlag, "log", "set the `level` of logging.")
+	flag.Usage = func() {
+		cmdName := path.Base(os.Args[0])
+		w := flag.CommandLine.Output()
+		fmt.Fprintf(w, "Usage:\n  %s [FLAGS]... FILE\n\n", cmdName)
+		fmt.Fprintf(w, "Enumerates GitHub repositories between -start date and -end date, with -min-stars\n")
+		fmt.Fprintf(w, "or higher. Writes each repository URL on a separate line to FILE.\n")
+		fmt.Fprintf(w, "\nFlags:\n")
+		flag.PrintDefaults()
+	}
+}
+
+// searchWorker waits for a query on the queries channel, starts a search with that query using s
+// and returns each repository on the results channel.
+//
+// When the queries channel is closed it will call wg.Done() to signal that the worker has finished.
+func searchWorker(s *githubsearch.Searcher, wg *sync.WaitGroup, logger logr.Logger, queries, results chan string) {
+	defer wg.Done()
+	for q := range queries {
+		total := 0
+		err := s.ReposByStars(q, *minStarsFlag, *starOverlapFlag, func(repo string) {
+			results <- repo
+			total++
+		})
+		if err != nil {
+			// TODO: this error handling is not at all graceful, and hard to recover from.
```

Comment:
Could consider having an `err` field in the `result` struct. In case of failure, this field will get populated and the error handling can happen after the go routines finish.

Reply:
Thanks for the suggestion. I haven't yet figured out exactly what I want to do here.

I plan to address this is a future PR.